Add documentation for the RAND_DRBG API

mspncp opened the pull request:
As announced long time ago in my personal roadmap #4403, I would like to publish the RAND_DRBG (and RAND_POOL) functions as officially supported API for the OpenSSL 1.1.1 release. 

This pull request adds the necessary documentation.

##### Note

A documentation update for pull request #5547 will be submitted as separate pull request shortly.

##### Checklist

- [x] Documentation is added or updated

kroeckx commented on doc/man3/RAND_DRBG_new.pod:
```diff
@@ -0,0 +1,150 @@
+The random bytes are used as entropy input for a full reseeding
+(resp. reinstantiation) if the DRBG is instantiated
+(resp. uninstantiated or in an error state).
+A reseeding requires 16 bytes (128 bits) of randomness.
```
This is not the case anymore, it's now 32. We probably at least need to return the drbg->min_entropylen.

mspncp commented:
I currently see no way to make `doc-nits` happy with the callbacks. No matter whether I use the typedefs (`RAND_DRBG_get_entropy_fn`, etc.) or the callback names (`get_entropy` etc.), I always get "... is not public", "... missing from NAME section", or "... missing from SYNOPSIS".

@levitte, any ideas?


```
doc/man3/RAND_DRBG_set_callbacks.pod:1: RAND_DRBG_get_entropy_fn is not public
doc/man3/RAND_DRBG_set_callbacks.pod:1: RAND_DRBG_get_nonce_fn is not public
doc/man3/RAND_DRBG_set_callbacks.pod:1: RAND_DRBG_cleanup_nonce_fn is not public
doc/man3/RAND_DRBG_set_callbacks.pod:1: RAND_DRBG_cleanup_entropy_fn is not public
...
doc/man3/RAND_DRBG_set_callbacks.pod:1: get_entropy missing from NAME section
doc/man3/RAND_DRBG_set_callbacks.pod:1: cleanup_entropy missing from NAME section
doc/man3/RAND_DRBG_set_callbacks.pod:1: get_nonce missing from NAME section
doc/man3/RAND_DRBG_set_callbacks.pod:1: cleanup_nonce missing from NAME section
...
doc/man3/RAND_DRBG_set_callbacks.pod:1: RAND_DRBG_cleanup_entropy_fn missing from SYNOPSIS
doc/man3/RAND_DRBG_set_callbacks.pod:1: RAND_DRBG_cleanup_nonce_fn missing from SYNOPSIS
doc/man3/RAND_DRBG_set_callbacks.pod:1: RAND_DRBG_get_nonce_fn missing from SYNOPSIS
doc/man3/RAND_DRBG_set_callbacks.pod:1: RAND_DRBG_get_entropy_fn missing from SYNOPSIS
...
doc/man3/RAND_DRBG_set_callbacks.pod:1: get_entropy missing from NAME section
doc/man3/RAND_DRBG_set_callbacks.pod:1: cleanup_entropy missing from NAME section
doc/man3/RAND_DRBG_set_callbacks.pod:1: get_nonce missing from NAME section
doc/man3/RAND_DRBG_set_callbacks.pod:1: cleanup_nonce missing from NAME section
...
doc/man3/RAND_DRBG_set_callbacks.pod:1: RAND_DRBG_get_entropy_fn is not public
doc/man3/RAND_DRBG_set_callbacks.pod:1: RAND_DRBG_cleanup_nonce_fn is not public
doc/man3/RAND_DRBG_set_callbacks.pod:1: RAND_DRBG_get_nonce_fn is not public
doc/man3/RAND_DRBG_set_callbacks.pod:1: RAND_DRBG_cleanup_entropy_fn is not public

```

richsalz commented:
Look at util/private.num

kroeckx commented on doc/man3/RAND_DRBG_generate.pod:
```diff
@@ -0,0 +1,78 @@
+                        int prediction_resistance,
+                        const unsigned char *adin, size_t adinlen);
+
+ int RAND_DRBG_bytes(RAND_DRBG *drbg, unsigned char *out, size_t outlen);
```
This functions (and probably others) are documented in the source. That documentation should probably get removed in favor of this one.

mspncp replied:
I had already started removing the comment docs, but then I noticed that some of the comments contain internal documentation, wich can't be moved to the manual pages, like the fact that the `drbg->lock` needs to be taken if non-null. Also, there are a lot of internal functions which have comment docs. Tearing out the comment docs for the public functions but leaving the ones for the private functions would have destroyed the consistency of the source code documentation. So I hesitated to remove it. Just removed the big explanation parts which are now essentially in man7.

mspncp commented:
Kurt, its nice that you already show so much interest in the documentation, but currently, it's really not more than an early snapshot of the current state. So please don't dig too deep yet :wink:

mspncp commented:
> Look at util/private.num

@richsalz, `RAND_DRBG_get_entropy_fn` is a typedef'd prototype for a callback function, which is named `get_entropy` in this manpage. So neither `RAND_DRBG_get_entropy_fn`, nor `get_entropy` is a symbol which is exported from libcrypto.

richsalz commented:
You can put the function typedef, but there are some formatting restrictions.  See BIO_set_callback.pod for example

levitte commented on doc/man3/RAND_DRBG_set_callbacks.pod:
```diff
@@ -0,0 +1,106 @@
+
+ typedef void (*RAND_DRBG_cleanup_nonce_fn)(
+                      RAND_DRBG *drbg,
+                      unsigned char *out, size_t outlen);
```
`doc-nits` reacts on these four typedefs, because you need to add them to `util/private.num`.  `grep '^OSSL_STORE.*_fn util/private.num` shows you other similar examples.

richsalz replied:
Or more properly if they are public typedefs then they probably have to be formatted specially.  See the regexp in util/find-doc-nits.

mspncp replied:
Thanks, I will check it this evening.

mspncp commented:
I removed the WIP because most of the manual pages are ready for review (see checklist at the beginning of the pull request).

Removing WIP also means that I will stop force-pushing until the review is finished. Then everything will be squashed into a single commit.

mspncp commented:
@kroeckx I'm still working on `RAND_POOL_acquire_entropy.pod`. When it's finished, we still can decide whether we want to publish the RAND_POOL API or leave it private for 1.1.1.

kroeckx commented on doc/man3/RAND_DRBG_generate.pod:
```diff
@@ -0,0 +1,84 @@
+This additional data is mixed into the internal state of the random
+generator but does not contribute to the entropy count.
+The additional data can be omitted by setting B<adin>=NULL and
+B<adinlen>=0;
```
I guess I prefer "to NULL" instead of "= NULL". I don't know how other manpages do this.

mspncp replied:
Changed (two locations).

kroeckx commented on doc/man3/RAND_DRBG_new.pod:
```diff
@@ -0,0 +1,115 @@
+The optional B<flags> argument specifies a set of bit flags which can be
+joined using the | operator. Currently, the only flag is
+RAND_DRBG_FLAG_CTR_NO_DF, which disables the use of a the derivation function
+ctr_df. For an explanation, see [NIST SP 800-90A Rev. 1].
```
I wonder if we should make this flag available or not.

mspncp replied:
From the viewpoint that libcrypto is a general purpose crypto library, I would leave it available. The alternative would be to tear out the no_df code parts.

kroeckx replied:
Reading things again, the no_df code should use "full entropy" and we have no way to request full entropy or generate it ourself because we didn't implement 10.1.2 from SP 800-90C.

mspncp replied:
Assuming I have a hardware RNG device then I could replace the callbacks entirely and seed the DRBG using no_df. Same holds when I do external conditioning. So no_df is not useful for us at the moment but might be useful for others.

kroeckx replied:
You're right

kroeckx commented on doc/man3/RAND_DRBG_new.pod:
```diff
@@ -0,0 +1,115 @@
+
+RAND_DRBG_new() and RAND_DRBG_secure_new()
+create a new DRBG instance of the given B<type>, allocated from the heap resp.
+the secure heap.
```
Should we point to some documentation about the secure heap?

mspncp replied:
Done

kroeckx commented on doc/man3/RAND_DRBG_new.pod:
```diff
@@ -0,0 +1,115 @@
+=head1 HISTORY
+
+RAND_event() and RAND_screen() were deprecated in OpenSSL 1.1.0 and should
+not be used.
```
I'm confused why this is mentioned here.

mspncp replied:
#metoo ;-). It's a copy&paste relict that needs to be removed. The skeleton for RAND_DRBG_new.pod was created by copying from RAND_add.pod.

kroeckx commented on doc/man3/RAND_DRBG_reseed.pod:
```diff
@@ -0,0 +1,84 @@
+These functions are part of the OpenSSL L<RAND_DRBG(7)> API.
+
+RAND_DRBG_reseed()
+reseeds the given B<drbg>,  mixing in the specified additional data provided
```
Remove a space?

kroeckx replied:
There seem to be various places with a double space. It's fine after a ".". (I have no idea if we ever agreed on a style for that.)

mspncp replied:
I'll check the double spaces. As for the double spaces after the full stop: IIRC, it was Richards recommendation to avoid this question by always starting a new line after the end of a sentence. This is also more diff-friendly if a sentence needs to be changed.

kroeckx commented on doc/man3/RAND_DRBG_reseed.pod:
```diff
@@ -0,0 +1,84 @@
+RAND_DRBG_reseed()
+reseeds the given B<drbg>,  mixing in the specified additional data provided
+in the buffer B<adin> of length B<adinlen>.
+The additional data can be omitted by setting B<adin>=NULL and B<adinlen>=0.
```
Maybe it should mention something about that it gets it's data either from the parent drbg or it's default source?

kroeckx replied:
Or maybe it's already clear to people what reseed does

mspncp replied:
Done.

kroeckx commented on doc/man3/RAND_DRBG_reseed.pod:
```diff
@@ -0,0 +1,84 @@
+If B<interval> > 0, then the B<drbg> will reseed automatically whenever the
+elapsed time since its last reseeding exceeds the given reseed time interval.
+For safety, a reseeding will also occur if the clock has been reset to a
+smaller value.
```
I think such implementation details shouldn't be mentioned.

mspncp replied:
You mean the last sentence? I will remove it.

kroeckx commented on doc/man3/RAND_DRBG_set_callbacks.pod:
```diff
@@ -0,0 +1,107 @@
+randomness bytes as return value.
+
+If the callback fails to acquire at least B<entropy> bits of randomness,
+it must return a buffer length of 0.
```
Maybe it should that it can indicate an error by returning 0?

mspncp replied:
changed to:
```
If the callback fails to acquire at least B<entropy> bits of randomness,
it must indicate an error by returning a buffer length of 0.
```

kroeckx commented on doc/man3/RAND_DRBG_set_callbacks.pod:
```diff
@@ -0,0 +1,107 @@
+The B<get_nonce>() and B<cleanup_nonce>() callbacks are analogous to
+B<get_entropy>() and B<cleanup_entropy>() and are used by the B<drbg>
+to request random bits for the nonce.
+These two callbacks are optional, it is allowed to pass NULL function pointers.
```
I don't think it's allowed to set them to NULL? Things will just fail them. But it's optional to call this function, internal defaults will be used in that case.

kroeckx replied:
Oh, but nonce is optional, entropy not.

kroeckx replied:
And the nonce is only optional when we don't use the derivation function, we don't seem to be checking that.

kroeckx replied:
And we're not even providing a nonce ourself by default

kroeckx replied:
#5503

kroeckx commented on doc/man3/RAND_DRBG_set_callbacks.pod:
```diff
@@ -0,0 +1,107 @@
+Their parameter lists need to match the function prototypes above.
+
+Setting the callbacks is allowed only if the DRBG has not been initialized yet.
+Otherwise, the operation will fail.
```
So this means you can't actually change it for the 3 internal DRBGs because we try to instantiate them at the moment you try to get a pointer to them.

mspncp replied:
Good point. The best solution I see for this problem is to add the statement to the documenation that `RAND_DRBG_uninstantiate` needs to be called before changing settings.  (Doing an automatic uninstantiate in these 'configuration' functions does not seem to be a good idea.)

kroeckx commented on doc/man3/RAND_DRBG_set_callbacks.pod:
```diff
@@ -0,0 +1,107 @@
+
+It is important that B<cleanup_entropy>() and B<cleanup_nonce>() clear the buffer
+contents safely before freeing it, in order not to leave sensitive information
+about the DRBG's state in memory.
```
I wonder if we should say something about that the entropy and nonce functions may be the same function. 

It might also be useful to mention that it should make sure the same buffer is not reused until the clear function is called, and that the easiest way to do that is by allocating a dynamic buffer.

mspncp replied:
> I wonder if we should say something about that the entropy and nonce functions may be the same function.

I can add a statement that it is allowed to reuse the callbacks.

> It might also be useful to mention that it should make sure the same buffer is not reused until the clear function is called, and that the easiest way to do that is by allocating a dynamic buffer.

The documentation is already very clear about that. It never mentions  a static buffer. It says that the buffer should be allocated:
```
It is expected that the callback allocates and fills a random buffer of size
```

kroeckx commented on doc/man7/RAND_DRBG.pod:
```diff
@@ -0,0 +1,312 @@
+=head2 Disclaimer
+
+Unless you have very specific requirements for your random generator,
+you won't need got know and understand the details presented here.
```
You might want to rewrite that sentence

mspncp replied:
Does it sound better like this?
```
Unless you have very specific requirements for your random generator,
it is in general not necessary to utilize the RAND_DRBG API directly.
The usual way to obtain random bytes is to use L<RAND_bytes(3)> or
L<RAND_priv_bytes(3)>, see also L<RAND(7)>.
```

kroeckx commented on doc/man7/RAND_DRBG.pod:
```diff
@@ -0,0 +1,312 @@
+The random bytes are used as entropy input for a full reseeding
+(resp. reinstantiation) if the DRBG is instantiated
+(resp. uninstantiated or in an error state).
+A reseeding requires 16 bytes (128 bits) of randomness.
```
It requires 32 bytes now, you should probably say something that it depends on the strength of the DRBG.

kroeckx commented on doc/man7/RAND_DRBG.pod:
```diff
@@ -0,0 +1,312 @@
+the other instance the I<child> DRBG.
+
+This is called chaining. It is possible create chains of more than two DRBG
+in a row.
```
Maybe you should point to the RAND_DRBG_new() function here

kroeckx commented on doc/man7/RAND.pod:
```diff
@@ -0,0 +1,78 @@
+=item *
+
+Modify the default settings of the OpenSSL RAND method by modifying the security
+parameters of the underlying DRBG, which is described in detail in L<RAND_DRBG(7)>.
```
I don't see how you can modify the default settings.

FdaSilvaYY commented on doc/man7/RAND.pod:
```diff
@@ -0,0 +1,78 @@
+=head1 NAME
+
+RAND
+- the OpenSSL random genator
```
genator ?

FdaSilvaYY commented on doc/man7/RAND_DRBG.pod:
```diff
@@ -0,0 +1,312 @@
+
+You need to integrate a previously unsupported entropy source. If this is the
+case, please have a look at L<RAND_POOL_acquire_entropy(3)> and its
+implementations in rand_unix.c, rand_win.c, and rand_vms.c, respectivily.
```
s/respectively/respectively/

FdaSilvaYY commented on doc/man3/RAND_DRBG_reseed.pod:
```diff
@@ -0,0 +1,84 @@
+=head1 NOTES
+
+The default OpenSSL random generator is already set up for automatic reseeding,
+so in general it is not necessary to reseed it explicitely, or to modify
```
s/explicitely/explicitly

FdaSilvaYY commented on doc/man3/RAND_add.pod:
```diff
@@ -24,43 +24,43 @@ Deprecated:
 
-RAND_event() and RAND_screen() are equivalent to RAND_poll().
+RAND_event() and RAND_screen() are equivalent to RAND_poll() and exist
+for compatibilty reasons only. See HISTORY section below.
```
s/compatibilty/compatibility

FdaSilvaYY commented on doc/man7/RAND.pod:
```diff
@@ -0,0 +1,78 @@
+=head1 DESCRIPTION
+
+Random numbers are a vital part of cryptography, they are needed to provide
+unpredictablility for tasks like key generation, creating salts, and many more.
```
s/unpredictablility/unpredictability

FdaSilvaYY commented on doc/man7/RAND.pod:
```diff
@@ -0,0 +1,78 @@
+the deterministic random bit generator (DRBG) model as described in
+[NIST SP 800-90A Rev. 1]. The default random generator will initialize
+automatically on first use and will be fully functional without having
+to be initialized ('seeded') explicitely.
```
s/explicitely/explicitly/

FdaSilvaYY commented on doc/man7/RAND_DRBG.pod:
```diff
@@ -0,0 +1,312 @@
+
+The <master> DRBG is not used directly by the application, only for reseeding
+the two other two  DRBG instances. It reseeds itself by obtaining randomness
+either from os entropy  sources or by consuming randomnes which was added
```
s/randomnes/randomness/

FdaSilvaYY commented on doc/man7/RAND_DRBG.pod:
```diff
@@ -0,0 +1,312 @@
+the RAND_DRBG interface is I<not> thread-safe.
+
+All other DRBG instances don't support locking, because they are
+intendended to be used by a single thread. Instead of accessing a single
```
s/intendended/intended/

FdaSilvaYY commented on doc/man7/RAND_DRBG.pod:
```diff
@@ -0,0 +1,312 @@
+intendended to be used by a single thread. Instead of accessing a single
+DRBG instance concurrently from different threads, it is recommended to
+instantiate a separate DRBG instance per thread. Using the same shared
+DRBG (preferrably the public DRBG) as parent of DRBG instances on
```
s/preferrably/preferably/

FdaSilvaYY commented on doc/man7/RAND_DRBG.pod:
```diff
@@ -0,0 +1,312 @@
+Not all options are available on all platforms.
+The default option is --with-rand-seed=os, which is in general the best
+choice for the given operating system. So in general, it is not necessary
+to specify this option explicitely.
```
s/explicitly/explicitely/

FdaSilvaYY commented:
A few changes/typo deserves to be backport to  1.1.0.

mspncp commented:
> A few changes/typo deserves to be backport to 1.1.0.

@FdaSilvaYY which typos do you mean? I did not find any of them in 1.1.0. This here is mostly new documentation which will not be backported.

mspncp commented:
Squashed all micro-fixups resulting from Kurt's comments into a single fixup.

mspncp commented:
Added some documentation for prediction resistance and removed RAND_POOL documentation.

mspncp commented:
All documents are now ready for review.

mspncp commented:
@FdaSilvaYY thanks for reporting all the typos. I just pushed a second fixup because I noticed that I missed some of your corrections. I hope that I finally got them all.  I really should get myself a spell checker 😉.

kroeckx commented on doc/man7/RAND_DRBG.pod:
```diff
@@ -0,0 +1,315 @@
+by L<RAND_bytes(3)> and L<RAND_priv_bytes(3)>.
+
+Currently, there are three DRBG instances,
+the <master>, <public>, and <private> DRBG.
```
I'm wondering if we should document this or not. Your other pull request doesn't actually seem to make those 3 available, which might just be an oversight. And as long as they are not available, I see little point in making public documentation for it and prefer to keep it internal documentation.

mspncp replied:
That was indeed an oversight, thanks. Documentation for `RAND_DRBG_get0_*` has been added.

kroeckx commented on doc/man7/RAND_DRBG.pod:
```diff
@@ -0,0 +1,315 @@
+approved for providing prediction resistance.
+Since the default DRBG implementation does not have access to such an approved
+entropy source, a request for prediction resistance will currently always fail.
+In other words, prediction resistance is currently not supported yet by the DRBG.
```
It is supported, provided they set their own callbacks.

mspncp replied:
The remark does not say that it is unimplemented. It only say that it is not available, because there is no approved entropy source. This should be enough detail for now. For the upcomig FIPS evaluation some things will have to be specified more precisely anyway.

kroeckx commented on doc/man3/RAND_DRBG_new.pod:
```diff
@@ -0,0 +1,117 @@
+Currently, all DRBG types are based on AES-CTR, so B<type> can be one of the
+following values: NID_aes_128_ctr, NID_aes_192_ctr, NID_aes_256_ctr.
+Before the DRBG can be used to generate random bits, it is necessary to set
+its type and to instantiate it.
```
Should we say that we always set the DRBG's strength to the maximum supported by the type?

kroeckx commented on doc/man7/RAND_DRBG.pod:
```diff
@@ -0,0 +1,315 @@
+
+=head1 SEE ALSO
+
+...WIP...
```
WIP

kroeckx commented:
So one thing I've been wondering about is the functions provide all the parameters we want to provide? I think the NIST standard has various optional features that we currently don't implement, and wonder if we ever will or not.

mspncp commented:
BTW: I did not forget the comments from your last review. I just did not have the time yet to address them.

mspncp commented:
The last commit already assumes that the public and private DRBG instance are thread local (#5547). The rest of the documentation still needs to be adapted, in particular `man7/RAND_DRBG.pod`,

kroeckx commented on doc/man3/RAND_DRBG_set_callbacks.pod:
```diff
@@ -0,0 +1,128 @@
+to request random bits for the nonce.
+These two callbacks are optional, it is allowed to pass NULL function pointers.
+It is also allowed to reuse the B<get_entropy>() and B<cleanup_entropy>()
+callbacks as B<get_nonce>() and B<cleanup_nonce>() callbacks.
```
The functions actualyl have a different type, and #5402 changes get_entropy to have an extra parameter, so this is no longer the case.

kroeckx replied:
If we go with the current patch in #5503, we might also want to document that here.

kroeckx commented on doc/man3/RAND_DRBG_set_callbacks.pod:
```diff
@@ -82,12 +82,14 @@ free the buffer allocated previously by get_entropy().
+The callbacks are analogous to get_entropy() and cleanup_entropy(),
+except for the missing prediction_resistance flag.
+
+If the derivation function is disabled, then no nonce is used for reseeding,
```
As far as I know, get_nonce is only used for instantiation, never for reseeding?

mspncp replied:
That's possible. I'll have to look it up again...

mspncp replied:
Corrected: 9f0207f2887b08b1b5c888b8471107280beb48d4

mspncp commented:
Given the fact that it's not clear yet whether #5547 will be accepted or not, the documentation is still a little undecided. I think, i will cleanup the commits in the next days and leave the difference between public and thread-local DRBGs in an extra commit which will either be squashed or discarded after the vote.

mspncp commented:
Ready for final review.

A documentation update for the thread local changes (pull request #5547) will be submitted as separate pull request shortly.

mspncp commented:
Documentation for #5547 is now available in #5629.

mspncp commented on doc/man7/RAND_DRBG.pod:
```diff
@@ -0,0 +1,315 @@
+This is in fact done automatically by L<RAND_DRBG_bytes(3)>.
+
+
+=head2 Configuring the Random Seed Source
```
Change request  by @levitte:

> This section is misplaced. However, INSTALL does need an update on this, and you could also add a NOTES.RANDOM...

mspncp replied:
Done, 47b5dd1dd860.

levitte commented on INSTALL:
```diff
@@ -208,12 +207,30 @@
+            egd:        Check for an entropy generating daemon.
+            rdcpu:      Use the RDSEED or RDRAND command if provided by the CPU.
+            librandom:  Use librandom (not implemented yet).
+            none:       Disable automatic seeding
```
Hmmm....  we don't usually outdent like this, see for example --prefix.

levitte commented on doc/man3/RAND_bytes.pod:
```diff
@@ -19,8 +19,10 @@ Deprecated:
 =head1 DESCRIPTION
 
+These functions are part of the OpenSSL L<RAND(7)> API.
+
```
Wait, what?  That was rather unusual...  and considering the names of the functions described, this comes as no surprise.  So this becomes a bit over the top.  I'd say it's sufficient to add `L<RAND(7)>` in the `SEE ALSO` section instead.

levitte replied:
that goes for *all* places where you've added this sentense...

levitte commented on doc/man3/RAND_set_rand_method.pod:
```diff
@@ -58,7 +60,7 @@ L<RAND_bytes(3)>, L<ENGINE_by_id(3)>
 =head1 COPYRIGHT
 
-Copyright 2000-2017 The OpenSSL Project Authors. All Rights Reserved.
+Copyright 2017-2018 The OpenSSL Project Authors. All Rights Reserved.
```
Why did you change 2000 to 2017?  If you `git log --stat --follow -C -M`, you will see that this page was, in fact, started in 2000.

mspncp replied:
Just noticed that the year was still wrong and pushed a fixup.

levitte commented on doc/man7/RAND.pod:
```diff
@@ -0,0 +1,78 @@
+=item *
+
+Modify the default settings of the OpenSSL RAND method by modifying the security
+parameters of the underlying DRBG, which is described in detail in L<RAND_DRBG(7)>.
```
@kroeckx, see #5632 (now merged)

levitte commented on doc/man7/RAND.pod:
```diff
@@ -0,0 +1,78 @@
+
+=head1 COPYRIGHT
+
+Copyright 2000-2018 The OpenSSL Project Authors. All Rights Reserved.
```
Should be just 2018

levitte commented on doc/man3/RAND_DRBG_get0_master.pod:
```diff
@@ -0,0 +1,69 @@
+L<RAND_DRBG_set_reseed_time_interval(3)>,
+L<RAND_DRBG_set_callbacks(3)>,
+L<RAND_DRBG(7)>,
+L<RAND(7)>
```
Why `RAND(7)`?  Or perhaps, why not in RAND_DRBG_generate.pod?

levitte commented on doc/man3/RAND_DRBG_set_ex_data.pod:
```diff
@@ -0,0 +1,67 @@
+
+L<CRYPTO_get_ex_data(3)>,
+L<CRYPTO_set_ex_data(3)>,
+L<CRYPTO_get_ex_new_index(3)>
```
Add `L<RAND_DRBG(7)>`?

levitte commented on INSTALL:
```diff
@@ -565,6 +592,10 @@
+                   DEVRANDOM       A comma separated list of strings that
+                                   specify pathes of random devices which are
+                                   searched in the given order for seeding the
+                                   random generator.
```
Wait, what?  I can see no such thing...

mspncp replied:
I added documentation about `DEVRANDOM`, because I noticed that documentation of `--with-rand-seed=devrandom` was inaccurate: It said
```
    devrandom:   Use the /dev/urandom device.
```
This is false, because the random device is chosen from a comma separated path list, which is `#define`d by the preprocessor constant `DEVRANDOM` in [e_os, line 29 ff](https://github.com/openssl/openssl/blob/master/e_os.h#L29). The `#define` is guarded by an `#ifndef`, so (in theory) it can be overwritten on the command line. My understanding was that it could be added to the configure arguments as `-DDEVRANDOM=...`. If that is wrong or too much level of detail, I'll remove it again.

Think over it twice, it is rather clumsy to have to modify both `--with-rand-seed` and `DEVRANDOM`. In particular, because of the necessity to quote the double quotes when passing `DEVRANDOM` on the commandline is rather messy.  (BTW, the same holds for `--with-rand=egd` and `DEVRANDOM_EGD`).

In fact, it would be much better to make `--with-rand-seed` more flexible, so one could configure the random sources directly using `--with-rand-seed` and not having to fiddle around with `DEVRANDOM`. A possible solution would be to provide the possibility for adding an optional path list as arguments to `devrandom` and `egd`. Here I chose `+` as path separator, because `,` already separates the seed methods. (The choice is arbitrary, maybe you have a better choice?)

```
--with-rand-seed=seed1[,seed2,....]

	devrandom[=/dev/devrandom+/dev/urandom+...]
	egd[=/dev/devrandom+/dev/urandom+...]	
```

OpenSSL would proved proper defaults, when `--with-rand-seed=os` is used. For example, it would imply

```
   --with-rand-seed="devrandom:/dev/urandom+/dev/random,/dev/srandom" on S390, and
   --with-rand-seed="devrandom:/dev/prandom+/dev/urandom+/dev/hwrng+/dev+random" otherwise.
```

But for such a change I would need your help (and/or help by @paulidale), because I'm not firm with perl and the build internals. Such a change comes practically too late for 1.1.1, so it would probably be better to leave it 1.2. (or not do it at all...)

Please let me know what your preferred solution for the moment ist, whether I should revert my mentioning of `DEVRANDOM` or how I should change its documentation. I will be happy to follow your advice.

mspncp replied:
Note: had to make some minor corrections to my previous post. It's better to read it online.

levitte replied:
But this part of `INSTALL` isn't about cpp macros, it's about "make variables" that can be passed to Configure.  In other words, what you're saying is that there would be a variable that you could control like this:

```
./config DEVRANDOM='/dev/random /dev/srandom'
```

There is no support for that variable.

mspncp replied:
Ok, I'll remove that hunk again. But now I have to leave, I'll do it later...

mspncp replied:
Or, alternatively:  Would it be a great effort for you to add such a make variable, which would create the `-DDEVRANDOM=...` compiler argument automatically? (And same for `DEVRANDOM_EGD`?)

levitte replied:
> Or, alternatively:  Would it be a great effort for you to add such a make variable, which would create the `-DDEVRANDOM=...` compiler argument automatically? (And same for `DEVRANDOM_EGD`?)

The diverse variables that are used to set CPP flags is already a complex enough matter, I see no reason why we should add to it, especially if the user can already do this:

```
./config -DDEVRANDOM='"/dev/random","/dev/srandom"'
```

or

```
./config CPPDEFINES='DEVRANDOM="/dev/random","/dev/srandom"'
```

In that case, I *much* prefer your idea of making the `--with-seed-source` argument more flexible. Trouble is that we're going into beta on tuesday, time's pressing...

mspncp replied:
Commit 6f82183662e8 reverts DEVRANDOM entry in make variable section.

levitte commented on doc/man3/CRYPTO_get_ex_new_index.pod:
```diff
@@ -52,8 +52,10 @@ The specific structures are:
     RSA
     ENGINE
     UI
-    UI_METHOD
```
I'd rather that stayed there, since `UI` and `UI_METHOD` belong together.

mspncp replied:
You are right: It does not make sense to order the names according to the numerical order of the `CRYPTO_EX_INDEX_*` constants in `<openssl/crypto.h>`. I added a fixup that reorders the names alphabetically (f8a576fef0fd).

levitte commented on doc/man3/RAND_set_rand_method.pod:
```diff
@@ -58,7 +57,7 @@ L<RAND_bytes(3)>, L<ENGINE_by_id(3)>
 
```
You might want to add `L<RAND(7)>`

levitte commented on doc/man3/RAND_DRBG_new.pod:
```diff
@@ -0,0 +1,115 @@
+L<OPENSSL_secure_zalloc(3)>,
+L<RAND_DRBG_generate(3)>,
+L<RAND_DRBG(7)>,
+L<RAND(7)>
```
Any particular reason why you have both `L<RAND_DRBG(7)>` and `L<RAND(7)>` here?

Note: this is just a question, doesn't mean you *have* to make a change...

mspncp replied:
I added a commit fixing all inconsistencies in 'SEE ALSO' sections (69783da897f8).

levitte commented on doc/man3/RAND_bytes.pod:
```diff
@@ -46,7 +46,7 @@ L<RAND_add(3)>
 
```
Add `L<RAND(7)>`

levitte replied:
This also goes for the RAND manual files you haven't touched yet...  `doc/man3/RAND_cleanup.pod`, `doc/man3/RAND_egd.pod` and `doc/man3/RAND_load_file.pod`

mspncp replied:
I added a commit fixing all inconsistencies in 'SEE ALSO' sections (69783da897f8).

mspncp commented:
Rebased without changes to resolve conflicts.

mspncp commented:
Added an entry to CHANGES file. Ready for final review (again).

mspncp commented:
Now since #5547 has been merged, #5629 has been merged with this pr.

mspncp commented:
This pull request needs to be rebased to resolve conflicts. Is it ok if I squash it while rebasing, or is anyone in the middle of reviewing it?

mspncp commented:
Rebased the branch (resolving conflicts) , squashed it into a single commit and reworded the commit message. The unsquashed branch (after rebase) was saved to [mspncp/openssl:pr-add-rand-drbg-documentation-unsquashed](https://github.com/mspncp/openssl/commits/pr-add-rand-drbg-documentation-unsquashed) for reference.

richsalz commented on INSTALL:
```diff
@@ -1164,10 +1191,22 @@
+ which can be used to specify a comma separated list of seed methods.
+ However in most cases OpenSSL will choose a suitable default method,
+ so it is not necessary to explicitely provide this option. Note also
+ that not all methods are available on all  platforms.
```
extra space here. :)

mspncp commented:
Thanks! Space has been removed.

I'll wait another day before I merge, just in case Kurt or Richard has any comments.

mspncp commented:
I should get into the habit of calling doc-nits myself ;-)

mspncp commented:
Travis builds are missing? Reopening to kick the CI bots...

mspncp commented:
@richsalz, apart from CI fixes I made a 'cosmetical' change (see f5e71e066d58): 
```
Renamed for consistency and clarity:
     public_drbg_thread_local_key  -> public_drbg
     private_drbg_thread_local_key -> private_drbg
     drbg_master                   -> master_drbg
```
Unless you object, I assume that your approval still holds.

richsalz commented:
still holds.

kroeckx commented on INSTALL:
```diff
@@ -209,12 +208,40 @@
+
+                   os:         Use a trusted operating system entropy source.
+                               This is the default method such an entropy source
+                               exists.
```
Probably needs an "if"

kroeckx commented on INSTALL:
```diff
@@ -1164,10 +1191,22 @@
+ and reseeding is disabled (--with-rand-seed=none) and it may be necessary
+ to install additional support software to obtain a random seed and reseed
+ the CSPRNG manually.  Please check out the manual pages for RAND_add(),
+ RAND_bytes(), RAND_egd(), and the FAQ for more information.
```
It might be useful to also look at the DRBG callbacks?

kroeckx commented on doc/man3/RAND_DRBG_new.pod:
```diff
@@ -0,0 +1,127 @@
+L<OPENSSL_zalloc(3)>,
+L<OPENSSL_secure_zalloc(3)>,
+L<RAND_DRBG_generate(3)>,
+L<RAND_DRBG(7)>
```
Should we mention something here about reseed intervals, or point to the defaults?

kroeckx commented on doc/man3/RAND_DRBG_reseed.pod:
```diff
@@ -0,0 +1,104 @@
+
+RAND_DRBG_reseed()
+reseeds the given B<drbg>, obtaining entropy input from its parent drbg
+(if it is chained) or from the trusted os entropy sources,
```
Maybe you should just say "from it's entropy source". Maybe you could mention the parent and the OS as examples. You can set other callbacks that get it from somewhere else, or you could have used RAND_add().

mspncp replied:
I added a fixup which rewords the description of entropy sources.

kroeckx commented on doc/man3/RAND_DRBG_set_callbacks.pod:
```diff
@@ -0,0 +1,147 @@
+see [NIST SP 800-90A Rev. 1]. Roughly speeking it can be omitted if the random
+source has "full entropy", i.e., contains 8 bits of entropy per byte.
+
+Even if a nonce is required but the B<get_nonce>() and B<cleanup_nonce>()
```
Remove the "but"?

mspncp replied:
Removed the "but" and reworded the get_nonce() paragraph a little.

kroeckx commented on doc/man7/RAND_DRBG.pod:
```diff
@@ -0,0 +1,285 @@
+DRBG instance concurrently from different threads, it is recommended to
+instantiate a separate DRBG instance per thread. Using the same shared
+DRBG (preferably the public DRBG) as parent of DRBG instances on
+different threads is safe.
```
I'm not sure what this is trying to say.

kroeckx commented on doc/man7/RAND_DRBG.pod:
```diff
@@ -0,0 +1,285 @@
+Calling RAND_poll() will always fail.
+
+RAND_add() needs to be called for initial seeding and periodic reseeding.
+At least 32 bytes (256 bits) of randomness have to be provided, otherwise
```
#5503 actually changes that to 48 bytes / 384 bit.

mspncp replied:
This one was missed. Just added an extra fixup right now.

mspncp commented:
Merged. Thanks for reviewing!
